sdk/eidolon_ai_sdk/security/msft_authorizer.py

```diff
@@ -0,0 +1,63 @@
+import os
+from typing import List, Optional, Any
+
+import httpx
+from jose import jwt
+from pydantic import BaseModel, Field
+
+from eidolon_ai_sdk.security.jwt_middleware import BaseJWTMiddleware
+from eidolon_ai_sdk.system.reference_model import Specable
+
+
+class MSFTJWTMiddlewareSpec(BaseModel):
+    jwks_url: str = Field(
+        "https://login.microsoftonline.com/common/discovery/keys",
+        description="The URL to fetch the JWKS from. Defaults to https://login.microsoftonline.com/common/discovery/keys",
+    )
+    audience: str = Field(
+        os.environ.get("AZURE_AD_CLIENT_ID"),
+        description="Your azure client or application ID. Defaults to the environment variable AZURE_AD_CLIENT_ID",
+    )
+    issuer_prefix: str = Field(
+        default="sts.windows.net",
+        description="The issuer prefix of the JWT. Defaults to sts.windows.net.  The tenant id will be appended to this value.  For example, sts.windows.net/your_tenant_id"
+    )
+    tenant_id: str = Field(
+        default=os.environ.get("AZURE_AD_TENANT_ID"), description="The tenant id of the JWT. Defaults to the environment variable AZURE_AD_TENANT_ID"
+    )
+
+
+class MSFTJWTMiddleware(BaseJWTMiddleware, Specable[MSFTJWTMiddlewareSpec]):
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.signing_keys = None
+        self.jwks_url = f"https://login.microsoftonline.com/{self.spec.tenant_id}/discovery/keys?appid={self.spec.audience}"
+
+    async def get_signing_keys(self):
+        if not self.signing_keys:
+            async with httpx.AsyncClient() as client:
+                resp = await client.get(self.jwks_url)
+                self.signing_keys = resp.json()["keys"]
+        return self.signing_keys
```

Comment:
get_signing_keys looks common, perhaps that should be pulled out to base BaseJWTMiddleware? you would need to a method get_jwks_url to BaseJWTMiddlewareSpec which each sub-spec implemented.

Reply:
See below conversation. I'm not sure yet.